ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/RelPlanParser.java

```diff
@@ -0,0 +1,477 @@
+import org.apache.calcite.sql.type.SqlTypeName;
+import org.apache.calcite.util.ImmutableBitSet;
+import org.apache.calcite.util.Pair;
+import org.apache.calcite.util.Util;
+import org.apache.hadoop.hive.common.TableName;
+import org.apache.hadoop.hive.ql.exec.ColumnInfo;
+import org.apache.hadoop.hive.ql.metadata.VirtualColumn;
+import org.apache.hadoop.hive.ql.optimizer.calcite.cost.HiveCostModel;
+import org.apache.hadoop.hive.ql.optimizer.calcite.cost.HiveDefaultCostModel;
+import org.apache.hadoop.hive.ql.optimizer.calcite.cost.HiveOnTezCostModel;
+import org.apache.hadoop.hive.ql.optimizer.calcite.reloperators.HiveGroupingID;
+import org.apache.hadoop.hive.ql.optimizer.calcite.reloperators.HiveRelNode;
+import org.apache.hadoop.hive.ql.optimizer.calcite.translator.TypeConverter;
+import org.apache.hadoop.hive.ql.parse.PrunedPartitionList;
+
+import java.io.IOException;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.nio.charset.Charset;
+import java.util.AbstractList;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+import static org.apache.commons.lang3.StringUtils.isNotBlank;
+
+/**
+ * Class responsible for parsing a given plan from a json file.
+ */
+public class RelPlanParser {
+  private static final TypeReference<LinkedHashMap<String, Object>> TYPE_REF =
+      new TypeReference<LinkedHashMap<String, Object>>() {
+      };
+
+  private final RelOptCluster cluster;
+  private final RelOptHiveTableFactory relOptHiveTableFactory;
+  private final Map<String, PrunedPartitionList> partitionCache;
+  private final HiveRelJson relJson = new HiveRelJson(null);
+  private final Map<String, RelNode> relMap = new LinkedHashMap<>();
+  private RelNode lastRel;
+
+  public RelPlanParser(RelOptCluster cluster,
+                       RelOptHiveTableFactory relOptHiveTableFactory,
+                       Map<String, PrunedPartitionList> partitionCache) {
+    this.cluster = cluster;
+    this.relOptHiveTableFactory = relOptHiveTableFactory;
+    this.partitionCache = partitionCache;
+  }
+
+  public RelNode parse(String json) throws IOException {
+    byte[] mapData = json.getBytes(Charset.defaultCharset());
+    return read(new String(mapData, Charset.defaultCharset()));
+  }
+
+  public RelNode read(String s) throws IOException {
+    JsonFactory jsonFactory = new JsonFactory();
+    Map<String, Object> o;
+    try (JsonParser parser = jsonFactory.createParser(s)) {
+      parser.nextToken();
+      assert parser.currentToken() == JsonToken.START_OBJECT;
+      parser.nextToken();
+      assert parser.currentToken() == JsonToken.FIELD_NAME;
+      parser.nextToken();
+      assert parser.currentToken() == JsonToken.VALUE_STRING;
```

Comment:
what happens in production code if one if these asserts fail? Should we throw exception instead?

Reply:
Adding a catch block for `AssertionError` and re-throwing it as an IOException.